qubrabench/algorithms/search.py

```diff
@@ -1,18 +1,98 @@
 """ This module provides a generic search interface that executes classically 
     and calculates the expected quantum query costs to a predicate function
 """
-
-from typing import Callable, Iterable, Optional, TypeVar
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from typing import Callable, Optional, TypeVar, Generic
 import numpy as np
 from ..stats import QueryStats
 
-
-__all__ = ["search"]
+__all__ = ["search", "SearchDomain"]
 
 
 E = TypeVar("E")
 
 
+class SearchDomain(ABC, Generic[E]):
+    """
+    Class used for search routines requiring elaborate sampling with replacement. To search against a predicate, please
+    refer to the function search_custom.
+    If your use-case involves simpler search spaces, e.g. ones that can be stored as a list in memory, please use the
+    function search for that.
+    """
+
+    # TODO: refer to 'simple' search function
+    # TODO: instead of size and num. solutions, ask for probability. Requires rewriting stat formulas!
+    @abstractmethod
+    def get_size(self) -> int:
+        pass
+
+    @abstractmethod
+    def get_number_of_solutions(self, key) -> float:
```

Comment:
perhaps make this `probability_of_sampling_solution` (or sth)? 
And later you can compute `T = get_size() * probprobability_of_sampling_solution()`